Documentation/AdditionalFunctionality/PushNotifications.md

```diff
@@ -0,0 +1,133 @@
+    case apns
+    case fcm
+}
+```
+
+The web service needs a specific route and `Component` to handle push notification registration. The following example shows how to add a `Device` to the currently authenticated `User`.
+
+```swift
+struct RegisterComponent: Component {
+    @Request(\.user) var user: User  // The logged in user
+
+    @Parameter var device: Device  // The `Device` instance in a request body
+
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    func handle() -> EventLoopFuture<Bool> {
+        notificationCenter.register(device, to: user);
+    }
+}
+```
+
+Furthermore, the `NotificationCenter` allows the removal and editing of `Device`s.
+
+## Sending Push Notifications
+
+`Notification`s are structured as follows:
+
+- **alert**: Displayed message on the device which includes title, subtitle, and body.
+- **payload**: Service specific notification settings like sound, badges, images, etc.
+- **data**: Background data.
+
+`Notification`s don't have to include an alert and can just send a silent push notification with background data.
+
+```swift
+public struct Notification {
+    public var alert: Alert
+    public var payload: Payload
+    public var data: Encodable
+}
+```
+
+Example of a `Component` that sends a push notification to an user:
+
+```swift
+struct SendNotification: Component {
+    @Request(\.user) var user: User
+
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    func handle() -> EventLoopFuture<Bool> {
+        notificationCenter.send(alert: .init(title: "Hello There 👋"), to: user)
+    }
+}
+```
+
+The `NotificationCenter` offers the following convenience methods to send push notifications:
+
+- Send to all registered devices
+- Send to a subscription
+- Send to all devices of a user
+- Send to multiple users
+- Send to a specific device
+- Send to a specific token
+
+## Unsolicited Push Notifications
+
+Events can trigger the sending of push notifications. This builds on top of `Component`s and follows the same approach as **[Service-Side Stream Components](../ComponentTypes/ServiceSideStream.md)** with `ObservableObject`s. `Component`s subscribe to `ObservableObject`s that are exposed to the environment. The handle method of the `Component` will be executed every time `@Published` properties of the `ObservableObject` change. `ObservableObject`s also can be scheduled as **[CronJobs](./CronJob.md)**.
+
+The following example shows a _WeatherService_ as a `CronJob` that fetches the weather every day and publishes it to its subscribers using the `@Published` property wrapper. The _AlertComponent_ listens to changes of the _WeatherService_ and sends out push notification based on the current weather. Every `Device` with the **subscription** _weatherSubscription_ will receive this push notification.
+
+```swift
+struct WeatherService: ObservedObject {
+    @Published var weather = "sunny"
+
+    handle() {
```

Comment:
A `func` is missing here. Also, I don't like the name `handle` as this name has a special place in Apodini and may confuse the reader. Or do you propose that Apodini should call the `handle` method on initialization of any `ObservedObject`?

Reply:
I also added a proposal for `CronJob`s. Therefore, the _WeatherService_ would be of type `CronJob` than `ObservedObject`. At startup this `CronJob` will be created and properties of it can be used in other `Component`s by injecting it with `@Environment`.